src/SymfonyCache/TaggableStore.php

```diff
@@ -0,0 +1,387 @@
+<?php
+
+/*
+ * This file is part of the FOSHttpCache package.
+ *
+ * (c) FriendsOfSymfony <http://friendsofsymfony.github.com/>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace FOS\HttpCache\SymfonyCache;
+
+use Psr\Cache\InvalidArgumentException;
+use Symfony\Component\Cache\Adapter\FilesystemAdapter;
+use Symfony\Component\Cache\Adapter\TagAwareAdapter;
+use Symfony\Component\HttpFoundation\HeaderBag;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpKernel\HttpCache\StoreInterface;
+use Symfony\Component\Lock\Exception\LockReleasingException;
+use Symfony\Component\Lock\Factory;
+use Symfony\Component\Lock\LockInterface;
+use Symfony\Component\Lock\Store\FlockStore;
+
+/**
+ * Implements a storage for Symfony's HttpCache that supports tagging.
+ *
+ * @author Yanick Witschi <yanick.witschi@terminal42.ch>
+ */
+class TaggableStore implements StoreInterface
+{
+    const NON_VARYING_KEY = 'non-varying';
+
+    /**
+     * @var string
+     */
+    private $purgeTagsHeader;
+
+    /**
+     * @var TagAwareAdapter
+     */
+    private $cache;
+
+    /**
+     * @var Factory
+     */
+    private $lockFactory;
+
+    /**
+     * @var LockInterface[]
+     */
+    private $locks = [];
+
+    /**
+     * @param string $cacheDir
+     */
+    public function __construct($cacheDir, $purgeTagsHeader = PurgeListener::DEFAULT_PURGE_TAGS_HEADER)
+    {
+        if (!class_exists(Factory::class)) {
+            throw new \RuntimeException('You need at least Symfony 3.4 to use the TaggableStore.');
+        }
+
+        $this->purgeTagsHeader = $purgeTagsHeader;
+        $this->cache = $cache = new TagAwareAdapter(new FilesystemAdapter('fos-http-cache', 0, $cacheDir));
```

Comment:
Adapter could be injected instead of built (if you inject it, you should inject the lock too, because the injected cache can be a shared cache like redis, and the Lock should be shared too)

Reply:
Both are not injectable on purpose. Otherwise it would be a fully PSR-6 compatible `Store` which was rejected in Symfony core as no other adapter except for file makes sense and might be dangerous using because of concurrency. See the discussion in https://github.com/FriendsOfSymfony/FOSHttpCache/issues/286 and the referenced PR in Symfony.

Reply:
You can TypeHint with TagAwareInterface ? What if someone want to use redis tombe share cache between multiple front?

Reply:
I guess you did not really understand: The users **must not** use any other adapter like redis etc. It **must** be forbidden. That's why it's hardcoded to the adapters that make sense.

Reply:
The point of Fabien is to not use PSR6 and simply use `$cached = include(...)`. (His comment give 3 good reasons).
As soon as you include cache component and use PSR6, I don't understand why adapters MUST be the Filesystem?

Reply:
Here's the comment: https://github.com/symfony/symfony/pull/20061#issuecomment-313339092

Quote: 

> The cache contains some PHP (when using ESI for instance) and storing PHP in anything else than a filesystem would mean eval()ing strings coming from Redis/Memcache/...,

So why would you allow anything else than `Filesystem`?

Reply:
Because here you use a cache adapter on top of filesystem and you must deal with the same issue.

Reply:
As soon as you start using PSR6, you are dealing with the same interface (same output, same evaluate things...) The only difference between adapter are performance vs scalability

Reply:
Hmm, I still don't get your point, sorry 😄  So trying to sum up: There was a pull request for Symfony core allowing to use any PSR-6 cache back end. It was rejected because it would allow to use Redis, Memcache etc. as cache back ends. Ergo: everything except for a filesystem cache does not make sense. However, for me this does not mean we cannot use the cache component. We just have to make sure nobody can use anything else than the `Filesystem` adapter.